### DISCUSSION TITLE
How do we test Presenter?

### DISCUSSION THREAD

**Discussion by stzn** [techexcellenceio/techexcellence#34] [Q&A]
Hi 

I saw some videos and blogs videos said “Test UseCases”.

https://www.youtube.com/watch?v=3wxiQB2-m2k&t=0s
https://www.youtube.com/watch?v=IZWLnn2fNko

But, how do we test Presenter?

I mean the Presenter layer in Clean Architecture.

![CleanArchitecture](https://user-images.githubusercontent.com/35151927/184247371-00b91349-4930-40d2-8654-2ccd336896de.jpg)

For example, I’m a mobile engineer. We have many presentation logics before showing displays.

1. Is it better to test them via E2E Level UseCase tests?
2. Is it better to test them separately as Unit Level tests?
3. Is there any better option?

In my opinion, 2 is better since sometimes there are many complex logics in Presenter, so it could be easier to assert the logics and find what's wrong quickly in independent tests. (although we need to be careful to frequent test code change.)

How do you think about it?

Thank you.

I'm not good at English, so sorry fo my poor English.

**Comment by PCreations** [comment]
Yes you could (and probably should) test your presenter separately in specific unit tests. The whole point of presenter is to have "dumb" views, so the presenter acts as an humble object, easily testable without the hassle of end to end testing.

**Reply by stzn** [reply]
Thank you for the answer.

> the presenter acts as an humble object

I learned it from "The Clean Architecture" book.

I believe that we need to be careful that Presenter defines behavior(what data is shown on views) to avoid depending on structure(how data is shown on views) and breaking tests frequently. Is it right? 

Also, is there any cautious point to implement Presenter?
